File: pokemonv2/src/main/java/com/pokemon/TypeChart.java
package com.pokemon;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class TypeChart {

    private static final Map<PokemonType, Map<PokemonType, Double>> chart = new HashMap<>();

    static {
        // AGUA: efectivo contra FUEGO, débil contra PLANTA
        addInteractions(PokemonType.AGUA, Map.of(
            PokemonType.FUEGO, 2.0,
            PokemonType.PLANTA, 0.5
        ));

        // FUEGO: efectivo contra PLANTA, débil contra AGUA
        addInteractions(PokemonType.FUEGO, Map.of(
            PokemonType.PLANTA, 2.0,
            PokemonType.AGUA, 0.5
        ));

        // PLANTA: efectivo contra AGUA, débil contra FUEGO
        addInteractions(PokemonType.PLANTA, Map.of(
            PokemonType.AGUA, 2.0,
            PokemonType.FUEGO, 0.5
        ));

        // NORMAL: sin ventajas ni desventajas específicas
        addInteractions(PokemonType.NORMAL, Collections.emptyMap());
    }

    private static void addInteractions(PokemonType type, Map<PokemonType, Double> interactions) {
        chart.put(type, new HashMap<>(interactions));
    }

    public static double getMultiplier(PokemonType attackerType, PokemonType defenderType) {
        return chart.getOrDefault(attackerType, Collections.emptyMap())
                    .getOrDefault(defenderType, 1.0);
    }
}

File: pokemonv2/frontend/src/constants/typeChart.js
// Tipos de Pokémon. Coincide 1:1 con com.pokemon.PokemonType.java
export const POKEMON_TYPES = {
    AGUA: 'AGUA',
    FUEGO: 'FUEGO',
    PLANTA: 'PLANTA',
    NORMAL: 'NORMAL',
};

// Tabla de efectividad. Replica com.pokemon.TypeChart.java
// chart[atacante][defensor] = multiplicador
const TYPE_CHART = {
    AGUA:   { FUEGO: 2.0, PLANTA: 0.5 },
    FUEGO:  { PLANTA: 2.0, AGUA: 0.5 },
    PLANTA: { AGUA: 2.0, FUEGO: 0.5 },
    NORMAL: {},
};

export const getTypeMultiplier = (attackerType, defenderType) => {
    return TYPE_CHART[attackerType]?.[defenderType] ?? 1.0;
};

export const getEffectivenessMessage = (multiplier) => {
    if (multiplier > 1) return '¡Es súper efectivo!';
    if (multiplier < 1) return 'No es muy efectivo...';
    return null;
};

// Traducción del tipo a su nombre visible en español
export const TYPE_LABELS = {
    AGUA: 'Agua',
    FUEGO: 'Fuego',
    PLANTA: 'Planta',
    NORMAL: 'Normal',
};
